<a href="https://colab.research.google.com/github/CarloRomeoGitHub/ML-exercises/blob/master/ML_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
#from google.colab import files
#!pip install -q kaggle


In [0]:
#uploaded = files.upload()


In [0]:
#IMPORT
import os
import sklearn
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import sklearn.utils
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score
import math


In [0]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

In [426]:
print(train.info(), test.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            332 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           417 non-null float

We can notice that test set miss a lot of entries in the column "cabin". It will be a little bit tricky to fill all missing features of test, considering that test values are less that mean values (in number) of train values.
First moment we will work only on train set.

In [427]:
train.describe()

PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

Sex and Embarked columns have string values.
Embarked means port of departed of specific passenger. I think that we don't really need this information now. The Fare column is embedded in Pclass.
Sex column has to be converted in int form. 
From info() and describe() we understand that Age miss some values (714 on 891).
I will use the Age column mean value: 29.

In [0]:
def preprocess(df):
  preprocess_df = df
  preprocess_df.drop(['PassengerId', 'Embarked', 'Fare', 'Ticket', 'Cabin', 'Name'], 1, inplace= True)
  preprocess_df['Age'].fillna('29', inplace= True)
  SexConv = []
  for x in preprocess_df['Sex']:
    if x == 'male': 
      SexConv.append('1')
    elif x == 'female':
      SexConv.append('0')
  preprocess_df['SexConv'] = SexConv
  preprocess_df.drop(['Sex'], 1, inplace=True)
  return preprocess_df

In [429]:

processDF = preprocess(train)
print(processDF)

#Scaling data
sc = StandardScaler()
x = train.drop(['Survived'], 1)
y = train['Survived']
x = pd.DataFrame(sc.fit_transform(x.values), index=x.index, columns= x.columns)


     Survived  Pclass Age  SibSp  Parch SexConv
0           0       3  22      1      0       1
1           1       1  38      1      0       0
2           1       3  26      0      0       0
3           1       1  35      1      0       0
4           0       3  35      0      0       1
..        ...     ...  ..    ...    ...     ...
886         0       2  27      0      0       1
887         1       1  19      0      0       0
888         0       3  29      1      2       0
889         1       1  26      0      0       1
890         0       3  32      0      0       1

[891 rows x 6 columns]


In [0]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state=42)

In [431]:
x_test

Pclass       Age     SibSp     Parch   SexConv
709  0.827377 -0.043103  0.432793  0.767630  0.737695
439 -0.369365  0.110771 -0.474545 -0.473674  0.737695
840  0.827377 -0.735532 -0.474545 -0.473674  0.737695
720 -0.369365 -1.812645 -0.474545  0.767630 -1.355574
39   0.827377 -1.197152  0.432793 -0.473674 -1.355574
..        ...       ...       ...       ...       ...
433  0.827377 -0.966342 -0.474545 -0.473674  0.737695
773  0.827377 -0.043103 -0.474545 -0.473674  0.737695
25   0.827377  0.649327  0.432793  5.732844 -1.355574
84  -0.369365 -0.966342 -0.474545 -0.473674 -1.355574
10   0.827377 -1.966518  0.432793  0.767630 -1.355574

[179 rows x 5 columns]

At first, let's try LogistRegression.

In [432]:
lg = LogisticRegression(solver= 'lbfgs', random_state=42)
lg.fit(x_train, y_train)
logistic_prediction = lg.predict(x_test)
score = accuracy_score(y_test, logistic_prediction)
#confusion_matrix(y_test, logistic_prediction, score=score)

f1 = f1_score(y_test, logistic_prediction)
print(score, f1)

0.8100558659217877 0.7671232876712328


Not so bad score! Now we are going to compare different linear models.
But first, let's confirm my thesis: do we really do not need of dropped columns?


In [433]:
train

Survived  Pclass Age  SibSp  Parch SexConv
0           0       3  22      1      0       1
1           1       1  38      1      0       0
2           1       3  26      0      0       0
3           1       1  35      1      0       0
4           0       3  35      0      0       1
..        ...     ...  ..    ...    ...     ...
886         0       2  27      0      0       1
887         1       1  19      0      0       0
888         0       3  29      1      2       0
889         1       1  26      0      0       1
890         0       3  32      0      0       1

[891 rows x 6 columns]

In [0]:
train = pd.read_csv('train.csv')

In [435]:
train

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
888          889         0       3  ...  23.4500   NaN         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[891 rows x 12 columns]

I am really sure about Name column can't help us. It doens't mean nothing related to our Machine Learning engine. I can't predict your death analyzing your name! Someone could say "we always can find a pattern!". This is correct, but in this case to use name I have to encode all names in dataset. 

In [0]:
train.drop(['Name'], 1, inplace= True)

In [437]:
train['Cabin'].isna().sum()

687

As we can see from previous line of code more than half Cabin column values are missing. Looking the specific entry we can notice about 5 or more classes. Therefore I decide to leave here that column.

In [0]:
train.drop(['Cabin'],1, inplace= True)

In [439]:
train.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(3)
memory usage: 69.7+ KB


In [0]:
def preprocess(df):
  preprocess_df = df
  preprocess_df['Embarked'].fillna('S', inplace= True)
  preprocess_df['Age'].fillna(29, inplace= True)
  preprocess_df.drop(['Ticket'], 1, inplace= True)
  EmbarkedCod = []
   
  c=0
  SexCod = []
  for i in preprocess_df['Sex']:
    if i == 'male':
      SexCod.append(0)
    elif i == 'female':
      SexCod.append(1)
  preprocess_df['SexCod'] = SexCod
  preprocess_df.drop(['Sex'], 1, inplace= True)
  for i in preprocess_df['Embarked']:
    if i == 'S':
      EmbarkedCod.append(1)
    elif i == 'Q':
      EmbarkedCod.append(2)
    elif i == 'C':
      EmbarkedCod.append(3)
    elif  i == 'nan':
      EmbarkedCod.append(1)

    
    print(i, )
    c =+ 1
  preprocess_df['EmbarkedCod'] = EmbarkedCod
  preprocess_df.drop(['Embarked'], 1, inplace= True)
  return preprocess_df

    

In [0]:
train_cod = preprocess(train)

With this For loop we have printed all values of Column Embarked with each index. So we can easly find nan values and hard remove them.

In [442]:
train_cod.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Age            891 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Fare           891 non-null float64
SexCod         891 non-null int64
EmbarkedCod    891 non-null int64
dtypes: float64(2), int64(7)
memory usage: 62.8 KB


In [0]:
x = train_cod.drop(['Survived'], 1)
y = train_cod['Survived']
x = pd.DataFrame(sc.fit_transform(x.values), index=x.index, columns= x.columns)

x_train, x_test, y_train, y_test = train_test_split(x, y, random_state = 42, test_size = 0.2)


In [444]:
lg = LogisticRegression(solver= 'lbfgs', random_state=42)
lg.fit(x_train, y_train)
logistic_prediction = lg.predict(x_test)
score = accuracy_score(y_test, logistic_prediction)
#confusion_matrix(y_test, logistic_prediction, score=score)

f1 = f1_score(y_test, logistic_prediction)
print(score, f1)

0.8100558659217877 0.7638888888888888


As we can see F1 score is a little bit lower than previous one. I think that it's better to use "clean data" instead of forcing training. For me Quality is better than quantity. 
Let's proceed with other models!